websocket/client.go

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) 2019-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+// Package websocket is a tiny websocket client purpose-built for load-testing.
+// It does not have any special features other than strictly what is needed.
+package websocket
+
+import (
+	"bytes"
+	"sync"
+
+	"github.com/gorilla/websocket"
+	"github.com/mattermost/mattermost-server/v5/mlog"
+	"github.com/mattermost/mattermost-server/v5/model"
+)
+
+const avgReadMsgSizeBytes = 1024
+
+// Client is the websocket client to perform all actions.
+type Client struct {
+	Url          string
+	EventChannel chan *model.WebSocketEvent
+
+	conn      *websocket.Conn
+	authToken string
+	sequence  int64
+	readWg    sync.WaitGroup
+	writeMut  sync.RWMutex
+}
+
+// NewClient4 constructs a new WebSocket client.
+func NewClient4(url, authToken string) (*Client, error) {
+	conn, _, err := websocket.DefaultDialer.Dial(url+model.API_URL_SUFFIX+"/websocket", nil)
+	if err != nil {
+		return nil, err
+	}
+
+	client := &Client{
+		Url:          url,
+		EventChannel: make(chan *model.WebSocketEvent, 100),
+
+		conn:      conn,
+		authToken: authToken,
+		sequence:  1,
+	}
+
+	client.readWg.Add(1)
+	go client.reader()
+
+	client.SendMessage(
+		model.WEBSOCKET_AUTHENTICATION_CHALLENGE,
+		map[string]interface{}{"token": authToken})
+
+	return client, nil
+}
+
+// Close closes the client.
+func (c *Client) Close() {
+	// If Close gets called concurrently during the time
+	// a connection-break happens, this will become a no-op.
+	c.conn.Close()
+	// Wait for reader to return.
+	// If the reader has already quit, this will just fall-through.
+	c.readWg.Wait()
+}
+
+func (c *Client) reader() {
+	defer func() {
+		close(c.EventChannel)
+		// Mark wg as Done.
+		c.readWg.Done()
+	}()
+
+	var buf bytes.Buffer
+	buf.Grow(avgReadMsgSizeBytes)
+
+	for {
+		// Reset buffer.
+		buf.Reset()
+		_, r, err := c.conn.NextReader()
+		if err != nil {
+			if !websocket.IsCloseError(err, websocket.CloseNormalClosure, websocket.CloseNoStatusReceived) {
+				// log error
+				mlog.Warn("error from conn.NextReader", mlog.Err(err))
+			}
+			return
+		}
+		// Use pre-allocated buffer.
+		_, err = buf.ReadFrom(r)
+		if err != nil {
+			if !websocket.IsCloseError(err, websocket.CloseNormalClosure, websocket.CloseNoStatusReceived) {
+				// log error
+				mlog.Warn("error from buf.ReadFrom", mlog.Err(err))
+			}
+			return
+		}
+
+		event := model.WebSocketEventFromJson(&buf)
+		if event == nil {
+			continue
+		}
+		if event.IsValid() {
+			c.EventChannel <- event
+		}
```

Comment:
So with the previous client I was able to get into a state where `c.EventChannel` was full and at the same time the `userentity.Disconnect()` was being called which meant this goroutine would leak forever (blocking on this line) even after the user was stopped/removed.  
I am wondering if this is still a possibility and since this piece of logic hasn't really changed I would say yes.
I think we could either wrap this in a `select` or maybe drain the event channel on shutdown. Thoughts?

Reply:
Yeah good point. This was a problem with the previous client also. We could do both I guess ?
- non-blocking send
- drain the event channel on disconnect

?

Reply:
I guess that if we do 1. then 2. is not really needed. But that also means we could potentially drop valid events instead of blocking until they are processed.
With 2. we would only drop when we are sure they are not going to be processed (since we are shutting down). 

It's a very extreme scenario which should never happen if the agent is running a sensible number of users so either solution I think is valid as long as we remove the possibility of the leak.

Reply:
went with 1.